### PR TITLE
AWS SDK v1 DynamoDB backend

### PR DESCRIPTION
**Pull request by akhomchenko**
Good evening.

Thank you a lot for awesome library. It looks really cool.

I wanted to use it for one of my projects and noticed lack of most common backend I use - AWS DynamoDB. At the time I've forked the code there was no `6.0` branch and so I based my code on `5.0` branch which had a lot of helpful abstract classes. Implementation was really simple.

Given that there is `6.0` that does not contain changes from `5.0` I have couple of questions:

* is `5.0` abandoned and will not be merged to `master`?
* if so - what is a best way to start with custom backend in `master`/`6.0`?
* if not - do you plan to release it soon?

General questions:

* are you interested in this contribution? It is WIP now but I will add more tests and docs to make it a little bit easier to use
* any special requirements? Current implementation passes tck test but maybe something you want me to double check
* do you have import, naming or formatting rules? Repository does not contain them. I am trying to follow your style (feel free to call out).

Notes:

* I can take a look on SDKv2 later as I haven't used it at all (unlike v1). I assume there are common things that can be extracted.

Thank you and happy holidays!

### CODE REVIEW

**Comment by vladimir-bukhtoyarov** [comment]
Hello @akhomchenko 

Do now worry. Efforts from ```5.0``` are not abandoned. ```5.0``` has been just renamed to ```7.0```. ```7.0``` by itself will be released after completion of documentation and javadocs. 

It was just a technical decision to use ```6.0``` branch instead of ```4.11```, because of backward-incompatible changes in the API(replaceConfiguration now accepts the second parameter) requires from us to increase the major version, but ```5.0``` version was already taken by Beta release without documentation. So I decided to never release ```5.0``` officially and just move all the efforts to the ```7.0``` and release it in 2021.

**Comment by vladimir-bukhtoyarov** [comment]
According to this request: I will merge it to ```7.0```` during the holidays. Thank you for your work.

**Comment by vladimir-bukhtoyarov** [comment]
@ akhomchenko hello,

I am going to review and merge this request. Did you finalize all planned works?

**Comment by akhomchenko** [comment]
Hey, sorry for the delay.

> Did you finalize all planned works?

I have update with documentation + extra safe guards locally. I will publish it by EOD for review. There are some things that I am not sure about and I want to hear your opinion as you are a lead maintainer.

I want to test it in personal AWS account during this weekend but that should not affect the review.

**Comment by vladimir-bukhtoyarov** [comment]
Ok. Just do not forget to remove WIP prefix when finishing.

**Comment by akhomchenko** [comment]
Updated. I left some comments and questions as TODOs and will add some as inline comments.

File: experimental/bucket4j-dynamodb-sdk-v1/pom.xml [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+            <name>DynamoDB Local Release Repository</name>
+            <url>https://s3-us-west-2.amazonaws.com/dynamodb-local/release</url>
+        </repository>
+    </repositories>
+    <dependencyManagement>
+        <dependencies>
+            <dependency>
+                <groupId>com.amazonaws</groupId>
+                <artifactId>aws-java-sdk-bom</artifactId>
+                <version>1.11.913</version>
```
Comment (akhomchenko) [inline experimental/bucket4j-dynamodb-sdk-v1/pom.xml:34]: Unfortunately I can not specify range for `dependencyManagement` so I am not sure what is the best path forward. My idea was to specify 1.11.913 <2.0 range to support updates.

File: experimental/bucket4j-dynamodb-sdk-v1/src/main/java/io/github/bucket4j/dynamodb/v1/BaseDynamoDBTransaction.java [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+        //
+        // docs: https://docs.aws.amazon.com/amazondynamodb/latest/APIReference/API_GetItem.html
+        // > If there is no matching item, GetItem does not return any data and there
+        // > will be no Item element in the response.
+        if (result == null || !result.containsKey(stateAttr)) {
+            return Optional.empty();
+        }
+
+        AttributeValue state = result.get(stateAttr);
+        if (state.getB() == null) {
```
Comment (akhomchenko) [inline experimental/bucket4j-dynamodb-sdk-v1/src/main/java/io/github/bucket4j/dynamodb/v1/BaseDynamoDBTransaction.java:52]: This is the first place I have a question for. So in my initial implementation I was returning empty optional here as I was not sure if throwing an exception from `get` is something that can be expected (method does not define if exception can be thrown if state is broken).

I switched to throwing as there is no possibility for code to succeed if state attribute has type different from `NULL` or `B`.

Comment (akhomchenko) [reply experimental/bucket4j-dynamodb-sdk-v1/src/main/java/io/github/bucket4j/dynamodb/v1/BaseDynamoDBTransaction.java:52]: This method does not support `NULL` value for `state` as from current logic `NULL` should never be written, but I can update code to support it.

File: experimental/bucket4j-dynamodb-sdk-v1/src/main/java/io/github/bucket4j/dynamodb/v1/StringDynamoDBTransaction.java [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+            throw new IllegalArgumentException("key is empty");
+        }
+        // TODO: bring in guava for Utf8#encodedLength ?
+        int length = key.getBytes(StandardCharsets.UTF_8).length;
+        if (length > MAX_SUPPORTED_LENGTH) {
+            throw new IllegalArgumentException(
+                    "key " + key + " has length of " + length + " bytes " +
+                    "while max allowed is " + MAX_SUPPORTED_LENGTH + " bytes"
+            );
+        }
```
Comment (akhomchenko) [inline experimental/bucket4j-dynamodb-sdk-v1/src/main/java/io/github/bucket4j/dynamodb/v1/StringDynamoDBTransaction.java:36]: I added this check in updated revision to provide better message. If you think that it is not worth the trouble and / or expensive to do I can drop it. DynamoDB will throw it's own exception.

**Comment by akhomchenko** [comment]
Oh wow, that was fast! Thank you for accepting my contribution! Could you please take a look on questions I left in this PR? I will do some follow-up fixed based on your feedback if necessary. I am mostly interested if `GenericEntry#getModifiedStateBytes` can ever return `null`. Thank you.

**Comment by vladimir-bukhtoyarov** [comment]
I will answer your questions at weekend

**Pull request merged**